### PR TITLE
Do cache warmer for vote using quic if enabled

### PR DESCRIPTION
**Pull request by lijunwangs**
#### Problem

In order to reduce vote latency when using QUIC, use cache warmer to warm up connections to leaders.  
#### Summary of Changes

Made cache warmer to support both vote and regular tpu.

Fixes #

### CODE REVIEW

File: core/src/tvu.rs [hunk 1]
```diff
@@ -417,6 +412,31 @@ impl Tvu {
...
+    cluster_info: &Arc<ClusterInfo>,
+    poh_recorder: &Arc<RwLock<PohRecorder>>,
+    exit: &Arc<AtomicBool>,
+) -> Option<WarmQuicCacheService> {
+    let tpu_connection_cache = connection_cache
+        .is_some_and(|cache| cache.use_quic())
+        .then(|| connection_cache.unwrap().clone());
+    let vote_connection_cache = vote_connection_cache
+        .use_quic()
+        .then_some(vote_connection_cache);
```
Comment (behzadnouri) [inline core/src/tvu.rs]: Below might be easier to read:
```rust
let tpu_connection_cache = connection_cache.filter(|cache| cache.use_quic()).cloned();
let vote_connection_cache = Some(vote_connection_cache).filter(|cache| cache.use_quic());
```

File: core/src/warm_quic_cache_service.rs [hunk 2]
```diff
@@ -26,13 +31,47 @@ const CACHE_OFFSET_SLOT: i64 = 100;
 const CACHE_JITTER_SLOT: i64 = 20;
 
 impl WarmQuicCacheService {
+    pub fn warmup_connection(
+        cache: Option<&Arc<ConnectionCache>>,
```
Comment (behzadnouri) [inline core/src/warm_quic_cache_service.rs]: We don't need `Arc` here, so this type can be:
```rust
cache: Option<&ConnectionCache>`,
```
then the calling method will use `.as_deref`:
```rust
Self::warmup_connection(
    tpu_connection_cache.as_deref(),
    ...
```

File: core/src/warm_quic_cache_service.rs [hunk 3]
```diff
@@ -26,13 +31,47 @@ const CACHE_OFFSET_SLOT: i64 = 100;
 const CACHE_JITTER_SLOT: i64 = 20;
 
 impl WarmQuicCacheService {
+    pub fn warmup_connection(
```
Comment (behzadnouri) [inline core/src/warm_quic_cache_service.rs]: Doesn't need to be `pub`.
Also could be a free function.

File: core/src/warm_quic_cache_service.rs [hunk 4]
```diff
@@ -26,13 +31,47 @@ const CACHE_OFFSET_SLOT: i64 = 100;
...
+        cache: Option<&Arc<ConnectionCache>>,
+        cluster_info: &ClusterInfo,
+        leader_pubkey: &Pubkey,
+        contact_info_selector: impl Fn(&ContactInfo) -> Result<SocketAddr, Error>,
+        log_context: &str,
```
Comment (behzadnouri) [inline core/src/warm_quic_cache_service.rs:39]:
```rust
log_context: &'static str,
```

File: core/src/warm_quic_cache_service.rs [hunk 5]
```diff
@@ -26,13 +31,47 @@ const CACHE_OFFSET_SLOT: i64 = 100;
...
+        if let Some(connection_cache) = cache {
+            if let Some(Ok(addr)) =
+                cluster_info.lookup_contact_info(leader_pubkey, contact_info_selector)
+            {
+                let conn = connection_cache.get_connection(&addr);
+                if let Err(err) = conn.send_data(&[]) {
+                    warn!(
+                        "Failed to warmup QUIC connection to the leader {:?} at {:?}, Context: {}, Error: {:?}",
+                        leader_pubkey, addr, log_context, err
+                    );
```
Comment (behzadnouri) [inline core/src/warm_quic_cache_service.rs]: https://rust-lang.github.io/rust-clippy/master/index.html#uninlined_format_args

File: core/src/warm_quic_cache_service.rs [hunk 6]
```diff
@@ -26,13 +31,47 @@ const CACHE_OFFSET_SLOT: i64 = 100;
...
+                || tpu_connection_cache
+                    .as_ref()
+                    .is_some_and(|cache| cache.use_quic())
+        );
+        assert!(
+            vote_connection_cache.is_none()
+                || vote_connection_cache
+                    .as_ref()
+                    .is_some_and(|cache| cache.use_quic())
+        );
```
Comment (behzadnouri) [inline core/src/warm_quic_cache_service.rs]:
```rust
assert!(matches!(
    tpu_connection_cache.as_deref(),
    None | Some(ConnectionCache::Quic(_))
));
assert!(matches!(
    vote_connection_cache.as_deref(),
    None | Some(ConnectionCache::Quic(_))
));
```

File: core/src/warm_quic_cache_service.rs [hunk 7]
```diff
@@ -48,20 +87,22 @@ impl WarmQuicCacheService {
...
-                                        "Failed to warmup QUIC connection to the leader {:?}, \
-                                         Error {:?}",
-                                        leader_pubkey, err
-                                    );
-                                }
-                            }
+                            // Warm cache for regular transactions
+                            Self::warmup_connection(
+                                tpu_connection_cache.as_ref(),
+                                cluster_info.as_ref(),
```
Comment (behzadnouri) [inline core/src/warm_quic_cache_service.rs]:
```rust
&cluster_info,
```

File: core/src/warm_quic_cache_service.rs [hunk 8]
```diff
@@ -48,20 +87,22 @@ impl WarmQuicCacheService {
...
+                                &leader_pubkey,
+                                |node| node.tpu(Protocol::QUIC),
+                                "tpu",
+                            );
+                            // Warm cache for vote
+                            Self::warmup_connection(
+                                vote_connection_cache.as_ref(),
+                                cluster_info.as_ref(),
```
Comment (behzadnouri) [inline core/src/warm_quic_cache_service.rs]:
```rust
&cluster_info,
```

File: core/src/warm_quic_cache_service.rs (hunk 3 above)
Comment (lijunwangs) [reply core/src/warm_quic_cache_service.rs]: I removed pub, but keep it in the structure it is about warming connection so it is relevant

File: core/src/warm_quic_cache_service.rs (hunk 5 above)
Comment (lijunwangs) [reply core/src/warm_quic_cache_service.rs]: done

File: core/src/warm_quic_cache_service.rs (hunk 6 above)
Comment (lijunwangs) [reply core/src/warm_quic_cache_service.rs]: done

File: core/src/warm_quic_cache_service.rs (hunk 8 above)
Comment (lijunwangs) [reply core/src/warm_quic_cache_service.rs]: done

File: core/src/warm_quic_cache_service.rs (hunk 7 above)
Comment (lijunwangs) [reply core/src/warm_quic_cache_service.rs]: done

File: core/src/warm_quic_cache_service.rs (hunk 2 above)
Comment (lijunwangs) [reply core/src/warm_quic_cache_service.rs]: done

File: core/src/warm_quic_cache_service.rs (hunk 4 above)
Comment (lijunwangs) [reply core/src/warm_quic_cache_service.rs:39]: I kept using &str as more flexible? Any performance difference using &'static here?

**Review by behzadnouri — APPROVED** [review]
lgtm

**Comment by lijunwangs** [comment]
In a separate branch with added logging: showing the cache being warmed for both vote and tpu:

The validator is run with this argument: 

--vote-use-quic true

[2024-12-15T04:44:29.547606797Z INFO  solana_core::warm_quic_cache_service] Warmed QUIC connection to the leader FQtkSv6J6eU9gk9jetwUDwrT3tC6cFdzQhB8j2ZWrDqi at 10.138.0.184:8009, Context: tpu
[2024-12-15T04:44:29.547650757Z INFO  solana_core::warm_quic_cache_service] Warmed QUIC connection to the leader FQtkSv6J6eU9gk9jetwUDwrT3tC6cFdzQhB8j2ZWrDqi at 10.138.0.184:8006, Context: vote
[2024-12-15T04:44:32.349215723Z INFO  solana_core::warm_quic_cache_service] Warmed QUIC connection to the leader BtqTeeN1NnEqKNKztdVLxH2UHLDziD3wYM9L3kjZVsc6 at 10.138.0.136:8009, Context: tpu
[2024-12-15T04:44:32.349259143Z INFO  solana_core::warm_quic_cache_service] Warmed QUIC connection to the leader BtqTeeN1NnEqKNKztdVLxH2UHLDziD3wYM9L3kjZVsc6 at 10.138.0.136:8006, Context: vote
[2024-12-15T04:44:37.952134458Z INFO  solana_core::warm_quic_cache_service] Warmed QUIC connection to the leader GAf2vgbuk4PqbxvsJwb3DEkJbnrFYLA5eAQ4uB2GEw2J at 10.138.0.150:8009, Context: tpu
[2024-12-15T04:44:37.952170266Z INFO  solana_core::warm_quic_cache_service] Warmed QUIC connection to the leader GAf2vgbuk4PqbxvsJwb3DEkJbnrFYLA5eAQ4uB2GEw2J at 10.138.0.150:8006, Context: vote

**Pull request merged**